File: 0049-group-anagrams/0049-group-anagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        # use a sorted version of the str as a key in tracker
        tracker = {}
        
        for str in strs:
            key = "".join(sorted(str))
            if key in tracker:
                tracker[key].append(str)
            else:
                tracker[key] = [str]
        
        return tracker.values()

File: 0049-group-anagrams/0049-group-anagrams.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs) {
    const tracker = {}
    
    for (const str of strs) {
        const key = getKey(str)
        
        if (!tracker[key]) tracker[key] = [str]
        else tracker[key].push(str)
    }
    
    return Object.values(tracker)
}; 

// n = max length (str1, str2); Time: O(n); Space: O(n)
const getKey = str => {
    const count = Array(26).fill(0)
    
    for (let i = 0; i < str.length; i++) {
        const index = str.toLowerCase().charCodeAt(i) - 97
        count[index]++
    }
    
    return JSON.stringify(count);
}